### ISSUE TITLE
Common coordinate system for videos

### ISSUE THREAD

**Issue by samcunliffe** [SainsburyWellcomeCentre/WAZP#3]
Possible considerations: the camera position may move. 
For e.g. heatmaps will need some common reference/alignment solution.

**Comment by adamltyson** [comment]
FYI - https://github.com/BrancoLab/Common-Coordinate-Behaviour

**Comment by sfmig** [comment]
Yes actually, I used them in my thesis. We used them to transform the 3D reconstruction we collected with a depth-sensing phone to the coordinate system of the motion capture cameras. Would have to refresh it a bit but yeah, sounds like a good solution :)

I guess for now we can add that as an enhancement? On the webapp side, how would that work? We could have a 'coordinate system registration' tab that transforms the bodyparts' coordinates from DLC (defined in the pixel space of the image) to the coordinate system defined by the ArUco marker. 

Another option could be to do the coordinate system registration during data collection, somewhat similar to what we were planning for the event tagging in the future. So with the camera in the position it will be in the experiment, the researcher would scan the aruco marker, and the transformation matrix would be computed and added to the video metadata. Then in the webapp we would apply that transform to the pose data. I just did a quick search but Bonsai seems to have ArUco tools (see Fig 4D in the [Bonsai paper](https://www.frontiersin.org/articles/10.3389/fninf.2015.00007/full), top of page 6/17 [in the bioRxiv paper](https://www.biorxiv.org/content/10.1101/006791v1.full.pdf), and [this Bonsai package](https://www.nuget.org/packages/Bonsai.Aruco.Design/#readme-body-tab)).

I guess a further enhancement could be to add pose tracking with ArUco markers (so no DLC), as they do with the head of the mouse in the paper. It easily provides 3D pose but you do need to stick a marker on the animal (and quite a big one actually). Less important but maybe of interest.

**Comment by niksirbi** [comment]
As this solution was suggested to me by a bonsai developer, I'm sure bonsai will have the tools to deal with it during data acquisition.

But we have to discuss it with Sanna and T&T, since they are the ones who would have to implement it in the setups, and there might be limits to how much we can modify the Zoo enclosures.

Other than that, I agree with you that the simplest use-case is computing the transform from the video, and then applying it to the keypoint tracks (applying it to the whole video will result in unnecessary data duplication and processing).

I am a big fan of the idea, it will remove the necessity for manual steps to define coordinates, which is always a big plus (provided it works reliably).

**Comment by sfmig** [comment]
In my experience it works quite robustly yeah.

One drawback is that it does add an extra dependency (since you must use aruco markers to define the coordinate system). What if you want to analyse old data for which you didn't place aruco markers? For ultimate flexibility™️, we may still want to consider alternative ways of registering a common coordinate system. So maybe a simpler solution like the ones we discussed based on clicking and fisheye corrections etc, which we implement in the prototype / v1 and keep as a fallback in the long term package

**Comment by niksirbi** [comment]
I agree, we anyhow already have many videos without markers, so we definitely still need the other solutions as well. Plus it's nice if other groups can use our tools for unmarked data. Ultimately, there should be a few registration options (manual, ArUco, etc.)

But given the scale of the Zoo project, ArUco markers are definitely worth considering.

**Comment by niksirbi** [comment]
They can definitely be made out of metal as well. The Aeon project uses metal (I think) plates with ArUCo markers engraved on them (for their octagon setup), and they mentioned ordering these from a company. We can ask them when the time comes.
But for small animals, printing it on a sticker will do. Just need to ensure that it's perfectly straight (no bending) and absolutely fixed at a certain position in the setup. The patterns can be programmatically generated by `cv2` and exported to image files ready for printing.

**Comment by sfmig** [comment]
I think the aruco solution is very neat but probably should be part of v1. 

We did consider implementing a simpler solution as part of v0, and I think @niksirbi started some work on this (I think it is on the `frame-registration` branch?). I can't find an issue for it, so will open one for now and add it to v0.

**Comment by niksirbi** [comment]
That's right, I am working on it on that branch, and it's good that you opened a separate issue for it. Let's keep this one as a reminder for a more permanent solution (e.g. ARUCO)
